components/search/search.tsx

```diff
@@ -0,0 +1,391 @@
+        // when focus is released from the search box.
+        setTimeout((): void => {
+            if (keepInputFocussed) {
+                setKeepInputFocussed(false);
+            } else {
+                setFocussed(false);
+            }
+        }, 0);
+
+        updateHighlightedSearchHint(0);
+    };
+
+    const handleSearchHintSelection = (): void => {
+        setKeepInputFocussed(true);
+    };
+
+    const handleAddSearchTerm = (term: string): void => {
+        const pretextArray = searchTerms?.split(' ') || [];
+        pretextArray.pop();
+        pretextArray.push(term.toLowerCase());
+        actions.updateSearchTerms(pretextArray.join(' '));
+        updateHighlightedSearchHint(0);
+    };
+
+    const handleUpdateSearchTerms = (term: string): void => {
+        actions.updateSearchTerms(term);
+        setFocussed(true);
+    };
+
+    const handleChange = (e: ChangeEvent<HTMLInputElement>): void => {
+        const term = e.target.value;
+        actions.updateSearchTerms(term);
+    };
+
+    const updateHighlightedSearchHint = (indexDelta: number, changedViaKeyPress = false): void => {
+        let newIndex = highlightedSearchHintIndex + indexDelta;
+
+        switch (indexDelta) {
+        case 1:
+            // KEY.DOWN
+            // is it at the end of the list?
+            newIndex = newIndex === visibleSearchHintOptions.length ? 0 : newIndex;
+            break;
+        case -1:
+            // KEY.UP
+            // is it at the start of the list (or initial value)?
+            newIndex = newIndex < 0 ? visibleSearchHintOptions.length - 1 : newIndex;
+            break;
+        case 0:
+        default:
+            // reset the index (e.g. on blur)
+            newIndex = -1;
+        }
+
+        setHighlightedSearchHintIndex(newIndex);
+        setIndexChangedViaKeyPress(changedViaKeyPress);
+    };
+
+    const handleEnterKey = (e: ChangeEvent<HTMLInputElement>): void => {
+        // only prevent default-behaviour, when one of the conditions is true
+        // when both are false just submit the form (default behaviour) with
+        // `handleSubmit` function called from the `form`
+        if (indexChangedViaKeyPress) {
+            e.preventDefault();
+            handleAddSearchTerm(visibleSearchHintOptions[highlightedSearchHintIndex].searchTerm);
+            setKeepInputFocussed(true);
+        }
+
+        if (props.isMentionSearch) {
+            e.preventDefault();
+            actions.updateRhsState(RHSStates.SEARCH);
+        }
+    };
+
+    const handleSubmit = (e: FormEvent<HTMLFormElement>): void => {
+        e.preventDefault();
+
+        handleSearch().then(() => {
+            setKeepInputFocussed(false);
+            setFocussed(false);
+        });
+    };
+
+    const handleSearch = async (): Promise<void> => {
+        const terms = searchTerms.trim();
+
+        if (terms.length === 0) {
+            return;
+        }
+
+        if (terms.length) {
```

Comment:
if we already checked for 0 terms isn't this redundant? Aren't we meaning to check for more than 1 term?

Reply:
true ... will change!